pyccel/stdlib/math/pyc_math.f90

```diff
@@ -98,9 +99,9 @@ function pyc_lcm_4(a, b) result(lcm)
 
     implicit none
 
-    integer(kind=4), value :: a
-    integer(kind=4), value :: b
-    integer(kind=4)        :: lcm
+    integer(C_INT32_T), intent(in) :: a
+    integer(C_INT32_T), intent(in) :: b
+    integer(C_INT32_T)        :: lcm
```

Comment:
I think that passing `a` and `b` by value was OK. Why did you change this?

Reply:
I was trying to use pure, but then I realize that pure function can call only pure functions.
I removed pure keyword but I forget to restore args to be passed by value.

Reply:
But is this calling a non-pure function?

Reply:
yes, lcm call gcd which is not a pure function, gcd can't be pure because it calls non pure functions mod and xor.

Reply:
`mod` is pure, but `xor` apparently isn't. Pyccel translates python's `^` operator to the function `IEOR`. This seems to be pure.
GCC's doc on XOR does not recommend using it https://gcc.gnu.org/onlinedocs/gfortran/XOR.html

Reply:
Passing by value doesn't seem to work with all compilers in pure functions. I get the following error:
```
Variable ‘b’ can not appear in a variable definition context (assignment) at (1) in PURE procedure
```
with gfortran 7.5

Reply:
Yes, pure functions need to have arguments as intent(in).
I will replace xor with ieor and make the two functions gcd and lcm pure.

Reply:
I am pretty sure that pure functions accept both `value` and `intent(in)` arguments. Please check the standard...

Reply:
I found this link for the 2003 standard:
https://j3-fortran.org/doc/year/04/04-007.pdf
Section 12.6 talks about pure functions.
> C1266 The specification-part of a pure function subprogram shall specify that all its nonpointer dummy
> data objects have INTENT(IN).

> C1267 The specification-part of a pure subroutine subprogram shall specify the intents of all its non
> pointer dummy data objects.

In theory this means that you could use `value` if you do:
```
value, intent(in)
```
I can't find anything in the standard saying this is not allowed, however a simple test with gfortran 7.5 shows that this doesn't work (at least with this compiler). I get:
```
Dummy argument ‘a’ with INTENT(IN) in variable definition context (assignment) at (1)
```
As an argument cannot be both `intent(in)` and `value` this precludes the use of `value` for `pure` functions

Reply:
Please check a newer version of the standard.

Reply:
https://j3-fortran.org/doc/year/10/10-007.pdf, Section 12.7:
> The specification-part of a pure function subprogram shall specify that all its nonpointer dummy data objects have the INTENT (IN) or the VALUE attribute.

The `value` attribute is part of the C-Fortran interoperability interface, so it might be quite recent.

Reply:
Having investigated and discussed this further, gfortran 9.3 gives the same results. Intel accepts this syntax from ifort 17.0 onwards.
In any case the error is raised at the assignment not at the declaration. This means that we can use `value` if we keep the variables `x` and `y`.

Inefficiencies caused by this approach will be easily optimised out by the compiler. But it has the advantage of ensuring that the interface matches the one that would be generated by pyccel (which encounters the same problems).